### PR TITLE
 Integrate replays into the client

### PR DESCRIPTION
**Pull request by MacChaeger**
They look like this: http://prntscr.com/jgb81s
Also see the corresponding server PR: https://github.com/Zarel/Pokemon-Showdown/pull/4663

### CODE REVIEW

File: .htaccess [hunk 1]
```diff
@@ -63,6 +63,10 @@ RewriteRule ^battle-([a-z0-9]+-[0-9]{8})$ https://replay.pokemonshowdown.com/$1
 RewriteCond %{HTTP_HOST} ^play\.pokemonshowdown\.com$ [NC]
 RewriteRule ^battle-([a-z0-9]+)$ https://replay.pokemonshowdown.com/$1 [R=302,L]
 
+# Redirect replay room links to the replay server
+RewriteCond %{HTTP_HOST} ^play\.pokemonshowdown\.com$ [NC]
+RewriteRule ^replay-([a-zA-Z0-9]+-[0-9]+)$ https://replay.pokemonshowdown.com/$1 [R=302,L]
+
```
Comment (Zarel) [inline .htaccess]: I said not to do replay room links...

File: js/client-chat.js [hunk 2]
```diff
@@ -28,7 +28,7 @@
 			var name = app.user.get('name');
 			var userid = app.user.get('userid');
 			if (this.expired) {
-				this.$chatAdd.text(this.expired === true ? 'This room is expired' : this.expired);
+				this.$chatAdd.html(this.expired === true ? 'This room is expired' : this.expired);
```
Comment (Zarel) [inline js/client-chat.js]: Eek, where is the validation? This makes me uncomfortable...

File: js/client.js [hunk 3]
```diff
@@ -2597,7 +2597,7 @@
 		initialize: function (data) {
 			var buf = '';
 			buf = '<p>Your replay has been uploaded! It\'s available at:</p>';
-			buf += '<p><a href="http://replay.pokemonshowdown.com/' + data.id + '" target="_blank">http://replay.pokemonshowdown.com/' + data.id + '</a></p>';
+			buf += '<p><a href="http://play.pokemonshowdown.com/replay' + data.id.slice(6) + '" target="_blank">http://play.pokemonshowdown.com/replay' + data.id.slice(6) + '</a></p>';
```
Comment (Zarel) [inline js/client.js]: ...no

File: js/client-chat.js (hunk 2 above)
Comment (MacChaeger) [reply js/client-chat.js]: Where's the validation for any of them?

Comment (Zarel) [reply js/client-chat.js]: `.text` doesn't need validation, `.html` is an XSS vector. We generally run `sanitizeHTML` or `escapeHTML` on any untrusted data we receive, before embedding it in HTML.

**Comment by Zarel** [comment]
I just need to update documentation. The client now requires a compile step: Run `node build` to update it after each change.

**Comment by MacChaeger** [comment]
When i run `node build` I get this:
`Error: ENOENT: no such file or directory, open 'C:\Users\msaim\OneDrive\Documents\GitHub\Pokemon-Showdown-Client\config\config.js'
    at Object.fs.openSync (fs.js:648:18)
    at Object.fs.readFileSync (fs.js:553:33)
    at Object.<anonymous> (C:\Users\[redacted]\OneDrive\Documents\GitHub\Pokemon-Showdown-Client\build-tools\update:31:20)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Function.Module.runMain (module.js:653:10)
    at startup (bootstrap_node.js:187:16)
child_process.js:641
    throw err;
    ^

Error: Command failed: node ./build-tools/update
    at checkExecSyncError (child_process.js:598:13)
    at execSync (child_process.js:638:13)
    at Object.<anonymous> (C:\Users\[redacted]\OneDrive\Documents\GitHub\Pokemon-Showdown-Client\build:35:1)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Function.Module.runMain (module.js:653:10)
    at startup (bootstrap_node.js:187:16)`

**Comment by Zarel** [comment]
The problem is that XHRs have permission issues. You're not actually allowed to make HTTP requests to other domains, except in very specific situations.

**Comment by Zarel** [comment]
I think a recent Chrome update prevented all HTTP requests, but I think that was reverted.

**Comment by Zarel** [comment]
Oh, I see. I enabled CORS. See if this works for you now.

**Comment by MacChaeger** [comment]
Nope, even after updating the branch with most recent changes and rebuilding, I still get exactly the same 404 error.

File: lib/dispatcher.lib.php [hunk 4]
```diff
@@ -369,6 +369,14 @@ public function uploadreplay($dispatcher, &$reqData, &$out) {
 		die($GLOBALS['Replays']->upload($reqData));
 	}
 
+	public function fetchreplay($dispatcher, &$reqData, &$out) {
```
Comment (Zarel) [inline lib/dispatcher.lib.php]: ...lol

I made a CORS change so you wouldn't have to roundtrip through `dispatcher` anymore.

File: js/client.js [hunk 5]
```diff
@@ -843,10 +843,38 @@
 						if (data === 'nonexistent' && Config.server.id && roomid.slice(0, 7) === 'battle-') {
 							var replayid = roomid.slice(7);
 							if (Config.server.id !== 'showdown') replayid = Config.server.id + '-' + replayid;
+							var replayFound = false;
 							var replayLink = 'http://replay.pokemonshowdown.com/' + replayid;
-							errormessage += '\n\nYou might want to try the replay: ' + replayLink;
+							var jqxhr = $.post(app.user.getActionPHP(), {
+								act: 'fetchreplay',
+								replaylogurl: replayLink + '.log'
```
Comment (Zarel) [inline js/client.js]: This should just be `$.get(replayLink + '.log')`

**Comment by MacChaeger** [comment]
>:/
I didn't know what a CORS change meant and you didn't say to change anything before...

**Comment by MacChaeger** [comment]
Well, I'm actually getting the log now, but I'm still not getting the `jqXHR` object that `$.get` is supposed to return according to its documentation which means I can't properly display error popups, and even with all the code related to that commented out, I get this crash:
>`Uncaught TypeError: this.callback is not a function
    at child.submit (client.js:2607)
    at child.dispatchSubmit (client.js:2281)
    at HTMLFormElement.wrapper (lodash.core.js:1157)
    at HTMLDivElement.dispatch (jquery-2.1.4.min.js:3)
    at HTMLDivElement.r.handle (jquery-2.1.4.min.js:3)`

**Comment by MacChaeger** [comment]
I mean I _guess_ I could make the request not be asynchronous.

**Comment by Zarel** [comment]
No you can't? I'm nearly sure you can't make a synchronous request; there are zero relevant APIs for that.

**Comment by Zarel** [comment]
Oh, it's because it's a cross-origin request, so PS replaces `$.get` and `$.post` with custom functions. I actually don't know why we don't just use CORS. I guess because it's easier than supporting XDomainRequest; at the time, IE9 support was probably important to us.

You can just use `$.ajax` to get at the native jQuery function.

```
var jqxhr = $.ajax('https://replay.pokemonshowdown.com/gen7randombattle-745860147.log');
```

is working for me.

File: js/client.js [hunk 6]
```diff
@@ -843,10 +843,27 @@
...
+								if (replay) {
+									var title = Tools.escapeHTML(replay.p1) + ' vs. ' + Tools.escapeHTML(replay.p2);
+									app.receive('>battle-' + replayid + '\n|init|battle\n|title|' + title + '\n' + replay.log);
+									app.receive('>battle-' + replayid + '\n|expire|<a href=' + replayLink + ' target="_blank">Open replay in new tab</a>');
+									replayFound = true;
+								} else {
+									errormessage += '\n\nResponse received, but no data.';
+								}
+							}).fail(function () {
+								errormessage += '\n\nYou might want to try the replay: ' + replayLink;
```
Comment (Zarel) [inline js/client.js]: This is a weird error message to give when you _know_ the replay doesn't exist...

File: js/client.js [hunk 7]
```diff
@@ -843,10 +843,27 @@
...
+							}).always(function () {
```
Comment (Zarel) [inline js/client.js]: I don't think there's a need for an `always` function; just handle the two conditions separately.

File: js/client.js [hunk 8]
```diff
@@ -843,10 +843,27 @@
 						if (data === 'nonexistent' && Config.server.id && roomid.slice(0, 7) === 'battle-') {
```
Comment (Zarel) [inline js/client.js]: Also, if you're going to do this, don't do `unjoinRoom` first it – i.e. don't close and reopen the room.

File: js/client-topbar.js [hunk 9]
```diff
@@ -90,7 +90,14 @@
 				return buf + ' aria-label="Join chatroom"><i class="fa fa-plus" style="margin:7px auto -6px auto"></i> <span>&nbsp;</span></a></li>';
 			case 'battle':
 				var name = Tools.escapeHTML(room.title);
-				var formatid = id.substr(7).split('-')[0];
+				var name = Tools.escapeHTML(room.title);
```
Comment (Zarel) [inline js/client-topbar.js]: This is clearly a bug.

Comment (MacChaeger) [reply js/client-topbar.js]: No idea how that happened.

File: js/client.js [hunk 10]
```diff
@@ -828,6 +828,29 @@
...
+							$.ajax(replayLink + '.json', {dataType: 'json'}).done(function (replay) {
+								if (replay) {
+									var title = Tools.escapeHTML(replay.p1) + ' vs. ' + Tools.escapeHTML(replay.p2);
+									app.receive('>battle-' + replayid + '\n|init|battle\n|title|' + title + '\n' + replay.log);
+									app.receive('>battle-' + replayid + '\n|expire|<a href=' + replayLink + ' target="_blank">Open replay in new tab</a>');
+								} else {
+									errormessage += '\n\nResponse received, but no data.';
+									app.addPopupMessage(errormessage);
+								}
+								isdeinit = false;
```
Comment (Zarel) [inline js/client.js]: This is asynchronous, it'll happen after `if (isdeinit)` below.

File: js/client.js [hunk 11]
```diff
@@ -828,6 +828,29 @@
...
+							}).fail(function () {
+								errormessage += "\n\nThe battle you're looking for has expired. Battles expire after 15 minutes of inactivity unless they're saved.\nIn the future, remember to click \"Save replay\" to save a replay permanently.";
+								app.addPopupMessage(errormessage);
+							});
+						} else {
+							this.addPopupMessage(errormessage);
+						}
```
Comment (Zarel) [inline js/client.js]: I would manually `return` after this line, and then manually call `unjoinRoom` in the failure handlers.

**Comment by Zarel** [comment]
Looks great!

**Pull request merged**